lib/util/exec/buildpack.ts

```diff
@@ -124,27 +141,51 @@ export async function resolveConstraint(
 
   const pkgReleases = await getPkgReleases(toolConfig);
   const releases = pkgReleases?.releases ?? [];
-  const versions = releases.map((r) => r.version);
-  const resolvedVersion = versions
-    .filter((v) =>
-      constraint ? versioning.matches(v, constraint) : versioning.isStable(v)
-    )
-    .pop();
 
-  if (resolvedVersion) {
-    logger.debug({ toolName, constraint, resolvedVersion }, 'Resolved version');
-    return resolvedVersion;
+  if (!releases.length) {
+    throw new Error('No tool releases found.');
   }
 
-  const latestVersion = versions.filter((v) => versioning.isStable(v)).pop();
-  if (!latestVersion) {
-    throw new Error('No tool releases found.');
+  const matchingReleases = releases.filter(
+    (r) => !constraint || versioning.matches(r.version, constraint)
+  );
+
+  const stableMatchingVersion = matchingReleases
+    .filter((r) => isStable(r.version, versioning, pkgReleases?.tags?.latest))
+    .pop()?.version;
+  if (stableMatchingVersion) {
+    logger.debug(
+      { toolName, constraint, resolvedVersion: stableMatchingVersion },
+      'Resolved stable matching version'
+    );
+    return stableMatchingVersion;
+  }
+
+  const unstableMatchingVersion = matchingReleases.pop()?.version;
+  if (unstableMatchingVersion) {
+    logger.debug(
+      { toolName, constraint, resolvedVersion: unstableMatchingVersion },
+      'Resolved unstable matching version'
+    );
+    return unstableMatchingVersion;
   }
+
+  const stableVersion = releases
+    .filter((r) => isStable(r.version, versioning, pkgReleases?.tags?.latest))
+    .pop()?.version;
+  if (stableVersion) {
+    logger.warn(
+      { toolName, constraint, stableVersion },
+      'No matching tool versions found for constraint - using latest stable version'
+    );
+  }
```

Comment:
isn't this already resolved by `stableMatchingVersion` ? Should we prefer stable before unstable?

Reply:
In this case if we don't get any matching versions at all then we resort to finding the highest stable version of all releases. Can you check my added tests and see if that makes sense to you? All code is covered

Reply:
I would expect this ?

```ts
 it('returns highest stable', async () => {
      expect(await resolveConstraint({ toolName: 'composer' })).toBe(
        '2.1.0'
      );
    });
```